File: webrtc/src/main/java/dev/onvoid/webrtc/RTCCertificatePEM.java
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.onvoid.webrtc;

/**
 * Represents a certificate used to authenticate WebRTC communications. This
 * class contains PEM strings of an RTCCertificate's private key and
 * certificate.
 *
 * @author Alex Andres
 */
public class RTCCertificatePEM {

	/**
	 * Indicates the time in milliseconds relative to the Unix epoch after which
	 * the certificate will be considered invalid.
	 */
	private final long expires;

	/**
	 * PEM string representation of the private key.
	 */
	private final String privateKey;

	/**
	 * PEM string representation of the certificate.
	 */
	private final String certificate;


	/**
	 * Creates a new RTCCertificatePEM instance with the specified PEM string
	 * representation of the private key and certificate.
	 *
	 * @param privateKey  Private key as PEM string.
	 * @param certificate Certificate as PEM string.
	 * @param expires     The expiration date of this certificate.
	 */
	public RTCCertificatePEM(String privateKey, String certificate, long expires) {
		this.privateKey = privateKey;
		this.certificate = certificate;
		this.expires = expires;
	}

	/**
	 * @return The PEM string representation of the private key.
	 */
	public String getPrivateKey() {
		return privateKey;
	}

	/**
	 * @return The PEM string representation of the certificate.
	 */
	public String getCertificate() {
		return certificate;
	}

	/**
	 * @return The expiration time in milliseconds of this certificate relative
	 * to the Unix epoch.
	 */
	public long getExpires() {
		return expires;
	}
}

File: webrtc-jni/src/main/cpp/src/rtc/RTCCertificatePEM.cpp
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "rtc/RTCCertificatePEM.h"
#include "JavaClasses.h"
#include "JavaObject.h"
#include "JavaString.h"
#include "JNI_WebRTC.h"

#include "rtc_base/ref_count.h"
#include "rtc_base/rtc_certificate_generator.h"

namespace jni
{
	namespace RTCCertificatePEM
	{
		JavaLocalRef<jobject> toJava(JNIEnv * env, const rtc::RTCCertificatePEM & certificate)
		{
			const auto javaClass = JavaClasses::get<JavaRTCCertificatePEMClass>(env);

			jobject obj = env->NewObject(javaClass->cls, javaClass->ctor,
				JavaString::toJava(env, certificate.private_key()).get(),
				JavaString::toJava(env, certificate.certificate()).get(),
				0
			);

			return JavaLocalRef<jobject>(env, obj);
		}

		rtc::RTCCertificatePEM toNative(JNIEnv * env, const JavaRef<jobject> & certificate)
		{
			const auto javaClass = JavaClasses::get<JavaRTCCertificatePEMClass>(env);

			JavaObject obj(env, certificate);

			return rtc::RTCCertificatePEM(
				JavaString::toNative(env, obj.getString(javaClass->privateKey)),
				JavaString::toNative(env, obj.getString(javaClass->certificate))
			);
		}

		JavaRTCCertificatePEMClass::JavaRTCCertificatePEMClass(JNIEnv * env)
		{
			cls = FindClass(env, PKG"RTCCertificatePEM");

			ctor = GetMethod(env, cls, "<init>", "(" STRING_SIG STRING_SIG "J)V");

			expires = GetFieldID(env, cls, "expires", "J");
			privateKey = GetFieldID(env, cls, "privateKey", STRING_SIG);
			certificate = GetFieldID(env, cls, "certificate", STRING_SIG);
		}
	}
}